In [ ]:
LAQ 3.1

In [ ]:
#Analysis of tumour neoantigen analysis using MuPeXi for WD13


import pandas as pd

# Load MuPeXi results (Assuming it's a TSV file)
df = pd.read_csv("C:/Users/dell/Downloads/WD13_Vcf.txt", sep="\t", skiprows=5)


# Apply filters for tumor neoantigens
neoantigens = df[(df["Norm_MHCrank_EL"] > 2) & (df["Mut_MHCrank_EL"] < 2)]

# Apply filters for strong tumor neoantigens
strong_neoantigens = neoantigens[neoantigens["Mut_MHCrank_EL"] < 0.5]

# Count unique neoantigen peptides per tumor
neoantigen_counts = neoantigens.groupby("Gene_Symbol").size()

# Count strong neoantigens per tumor
strong_neoantigen_counts = strong_neoantigens.groupby("Gene_Symbol").size()

# Print results
print("Total Tumor Neoantigens per Tumor Sample of WD13:")
print(neoantigen_counts)

print("\nTotal Strong Tumor Neoantigens per Tumor Sample of WD13:")
print(strong_neoantigen_counts)

print(f"Total Tumour Neoantigens of WD13: {len(neoantigens)}")
print(f"Total Strong Tumour Neoantigens of WD13: {len(strong_neoantigens)}")

# Save results to Excel
neoantigens.to_excel("Neoantigens_WD13.xlsx", index=False)
strong_neoantigens.to_excel("Strong_Neoantigens_WD13.xlsx", index=False)

LAQ 3-Part 1:
(1) How many tumour neoantigens can you find for this tumour? 366
(2) How many strong tumour neoantigens are there for this tumour? 37

In [2]:
import pandas as pd

# Load MuPeXi results (Assuming it's a TSV file)

df = pd.read_csv("C:/Users/dell/Downloads/WD06.mupexi.txt", sep="\t", skiprows=5)


# Apply filters for tumor neoantigens
neoantigens = df[(df["Norm_MHCrank_EL"] > 2) & (df["Mut_MHCrank_EL"] < 2)]

# Apply filters for strong tumor neoantigens
strong_neoantigens = neoantigens[neoantigens["Mut_MHCrank_EL"] < 0.5]

# Count unique neoantigen peptides per tumor
neoantigen_counts = neoantigens.groupby("Gene_ID").size()

# Count strong neoantigens per tumor
strong_neoantigen_counts = strong_neoantigens.groupby("Gene_ID").size()

# Print results
print("Total Tumor Neoantigens per Tumor Sample of WD6:")
print(neoantigen_counts)

print("\nTotal Strong Tumor Neoantigens per Tumor Sample of WD6:")
print(strong_neoantigen_counts)

print(f"Total Tumour Neoantigens of WD6: {len(neoantigens)}")
print(f"Total Strong Tumour Neoantigens of WD6: {len(strong_neoantigens)}")

# Save results to Excel
neoantigens.to_excel("Neoantigens_WD6.xlsx", index=False)
strong_neoantigens.to_excel("Strong_Neoantigens_WD6.xlsx", index=False)

Total Tumor Neoantigens per Tumor Sample of WD6:
Gene_ID
ENSG00000004399    2
ENSG00000007174    1
ENSG00000008277    1
ENSG00000008513    2
ENSG00000009307    2
                  ..
ENSG00000279012    6
ENSG00000279219    2
ENSG00000279301    2
ENSG00000283186    3
ENSG00000283652    1
Length: 305, dtype: int64

Total Strong Tumor Neoantigens per Tumor Sample of WD6:
Gene_ID
ENSG00000008277    1
ENSG00000008513    1
ENSG00000071189    2
ENSG00000087338    1
ENSG00000088179    1
ENSG00000090512    1
ENSG00000096060    1
ENSG00000100281    2
ENSG00000100784    1
ENSG00000101311    2
ENSG00000110427    2
ENSG00000113212    1
ENSG00000120549    1
ENSG00000123485    1
ENSG00000127481    1
ENSG00000128250    1
ENSG00000133124    1
ENSG00000136240    1
ENSG00000139613    1
ENSG00000139629    2
ENSG00000140463    1
ENSG00000142731    2
ENSG00000144476    1
ENSG00000149507    1
ENSG00000154864    2
ENSG00000156232    2
ENSG00000157856    1
ENSG00000159840    1
ENSG00000160145    1
ENSG00000160

LAQ 3-Part 1:
(1) How many tumour neoantigens can you find for this tumour? 570
(2) How many strong tumour neoantigens are there for this tumour? 66

In [3]:
import pandas as pd

# Load MuPeXi results (Assuming it's a TSV file)

df = pd.read_csv("C:/Users/dell/Downloads/WD14.mupexi.txt", sep="\t", skiprows=5)


# Apply filters for tumor neoantigens
neoantigens = df[(df["Norm_MHCrank_EL"] > 2) & (df["Mut_MHCrank_EL"] < 2)]

# Apply filters for strong tumor neoantigens
strong_neoantigens = neoantigens[neoantigens["Mut_MHCrank_EL"] < 0.5]

# Count unique neoantigen peptides per tumor
neoantigen_counts = neoantigens.groupby("Gene_Symbol").size()

# Count strong neoantigens per tumor
strong_neoantigen_counts = strong_neoantigens.groupby("Gene_Symbol").size()

# Print results
print("Total Tumor Neoantigens per Tumor Sample of WD14:")
print(neoantigen_counts)

print("\nTotal Strong Tumor Neoantigens per Tumor Sample of WD14:")
print(strong_neoantigen_counts)

print(f"Total Tumour Neoantigens of WD14: {len(neoantigens)}")
print(f"Total Strong Tumour Neoantigens of WD14: {len(strong_neoantigens)}")

# Save results to Excel
neoantigens.to_excel("Neoantigens_WD14.xlsx", index=False)
strong_neoantigens.to_excel("Strong_Neoantigens_WD14.xlsx", index=False)

Total Tumor Neoantigens per Tumor Sample of WD14:
Gene_Symbol
ANKRD16     1
AQP4        1
ASIC4       1
BGLAP       6
C1orf168    1
CALCOCO2    1
CAMSAP3     1
CCDC158     1
CEP295      1
DAG1        1
DNAH8       2
DPPA5       2
FBXL14      1
FBXO22      1
FLG2        1
GTF3C1      1
HAUS6       2
ITIH6       3
KBTBD8      1
KCNA2       1
LINGO1      1
MUC12       1
NBAS        3
NEDD4       1
NUP107      1
OR4D9       1
PCDH12      1
PKD1L1      2
PLCB3       2
RNF133      1
SAMD9L      1
SPTBN1      1
STK11IP     1
SYNE2       1
TMEM132C    2
TNPO2       2
UBAP1       1
UBE2Q2      6
ZNF714      4
ZNF780A     2
dtype: int64

Total Strong Tumor Neoantigens per Tumor Sample of WD14:
Gene_Symbol
BGLAP      1
CAMSAP3    1
DPPA5      1
HAUS6      1
NBAS       1
UBE2Q2     2
ZNF780A    1
dtype: int64
Total Tumour Neoantigens of WD14: 65
Total Strong Tumour Neoantigens of WD14: 8


LAQ 3-Part 1:
(1) How many tumour neoantigens can you find for this tumour? 65
(2) How many strong tumour neoantigens are there for this tumour? 8

In [4]:
import pandas as pd

# Load MuPeXi results (Assuming it's a TSV file)

df = pd.read_csv("C:/Users/dell/Downloads/WD22.mupexi.txt", sep="\t", skiprows=5)


# Apply filters for tumor neoantigens
neoantigens = df[(df["Norm_MHCrank_EL"] > 2) & (df["Mut_MHCrank_EL"] < 2)]

# Apply filters for strong tumor neoantigens
strong_neoantigens = neoantigens[neoantigens["Mut_MHCrank_EL"] < 0.5]

# Count unique neoantigen peptides per tumor
neoantigen_counts = neoantigens.groupby("Gene_Symbol").size()

# Count strong neoantigens per tumor
strong_neoantigen_counts = strong_neoantigens.groupby("Gene_Symbol").size()

# Print results
print("Total Tumor Neoantigens per Tumor Sample of WD22:")
print(neoantigen_counts)

print("\nTotal Strong Tumor Neoantigens per Tumor Sample of WD22:")
print(strong_neoantigen_counts)

print(f"Total Tumour Neoantigens of WD22: {len(neoantigens)}")
print(f"Total Strong Tumour Neoantigens of WD22: {len(strong_neoantigens)}")

# Save results to Excel
neoantigens.to_excel("Neoantigens_WD22.xlsx", index=False)
strong_neoantigens.to_excel("Strong_Neoantigens_WD22.xlsx", index=False)

print(df)

Total Tumor Neoantigens per Tumor Sample of WD22:
Gene_Symbol
ABCA1     1
ABCA12    1
ABCA3     2
ABRACL    1
ACAD8     1
         ..
ZFP42     1
ZNF214    4
ZNF534    1
ZNF541    1
ZNF692    1
Length: 256, dtype: int64

Total Strong Tumor Neoantigens per Tumor Sample of WD22:
Gene_Symbol
ARHGAP15     2
ASXL1        1
BCO2         1
BIN1         1
C1S          1
CASP12       1
CEP295       1
CNTLN        1
DGKD         1
DSEL         1
ELAVL4       1
EPB42        1
EPG5         4
FBXO21       1
FILIP1       1
FLRT2        1
GABRB3       1
GATA2        2
GLTSCR1L     1
GPR142       1
GRINA        1
ICOSLG       4
KCNG4        1
KIR2DL4      1
KLC1         1
LAMB4        3
MAP7D1       1
METTL23      1
MROH7        1
MYBPC2       2
MYO3A        1
NRXN1        1
OR13C3       1
OR1S1        1
OR5K3        1
PPP3CB       1
PSEN2        1
PTPRC        1
PXDN         1
RERG         1
RPH3A        1
SCML2        1
SERPINI2     2
SHE          1
SLC9C2       1
SPATA31E1    1
TACC2        1
TAS2R

LAQ 3-Part 1:
(1) How many tumour neoantigens can you find for this tumour?4945
(2) How many strong tumour neoantigens are there for this tumour70 88


In [5]:
df_06 = pd.read_excel("Strong_Neoantigens_WD6.xlsx", usecols=["Gene_Symbol", "Norm_peptide", "Mut_MHCrank_EL", "Norm_MHCrank_EL", "HLA_allele"])
df_13 = pd.read_excel("Strong_Neoantigens_WD13.xlsx", usecols=["Gene_Symbol", "Norm_peptide", "Mut_MHCrank_EL", "Norm_MHCrank_EL", "HLA_allele"])
df_14 = pd.read_excel("Strong_Neoantigens_WD14.xlsx", usecols=["Gene_Symbol", "Norm_peptide", "Mut_MHCrank_EL", "Norm_MHCrank_EL", "HLA_allele"])
df_22 = pd.read_excel("Strong_Neoantigens_WD22.xlsx", usecols=["Gene_Symbol", "Norm_peptide", "Mut_MHCrank_EL", "Norm_MHCrank_EL", "HLA_allele"])

df_06["Tumor Sample"] = "WD06"
df_13["Tumor Sample"] = "WD13"
df_14["Tumor Sample"] = "WD14"
df_22["Tumor Sample"] = "WD22"

concatenated_df = pd.concat([df_06, df_13, df_14, df_22], ignore_index=True)

concatenated_df.columns = ["HLA_allele", "Neoantigen Peptide", "Normal Rank Score", "Tumor Rank Score", "Affected Gene", "Tumor Sample"]

concatenated_df.to_csv("strong_tumor_neoantigens_all_samples.csv", index=False)

print(concatenated_df)

     HLA_allele Neoantigen Peptide  Normal Rank Score  Tumor Rank Score  \
0    HLA-A24:02         .........S             7.2538            0.0156   
1    HLA-A24:02          ........S             7.2642            0.0198   
2    HLA-A24:02          ........S             7.5499            0.0316   
3    HLA-A24:02          ........S             9.5501            0.0417   
4    HLA-A02:01          ........G            11.6109            0.0767   
..          ...                ...                ...               ...   
176  HLA-A02:01          ...V.....             2.1710            0.1317   
177  HLA-A24:02         .........L             2.1402            0.3818   
178  HLA-A24:02          .R.......             2.1333            0.1381   
179  HLA-A02:01          ..R......             2.0681            0.4954   
180  HLA-A24:02        ..........L             2.0317            0.4550   

    Affected Gene Tumor Sample  
0        SIGLEC12         WD06  
1        C1orf186         WD06  


LAQ 3- Part 3
Output of this part of the question are stored in the LAQ_3 -> Output_fies3

In [6]:

# Load the CSV data
shared = pd.read_csv("strong_tumor_neoantigens_all_samples.csv")


# Group by the 'Neoantigen Peptide' and count how many unique tumor samples have each neoantigen
shared_neoantigens = shared.groupby("Neoantigen Peptide")["Tumor Sample"].nunique()

print(shared_neoantigens)

# Filter for neoantigens that are present in more than one sample
shared_neoantigens = shared_neoantigens[shared_neoantigens > 1]

# Display the shared neoantigens
print(shared_neoantigens)

# Optional: You can also extract the full rows for the shared neoantigens
shared_neoantigen_rows = shared[shared["Neoantigen Peptide"].isin(shared_neoantigens.index)]
print(shared_neoantigen_rows)

# Save the shared neoantigens to a CSV (optional)
shared_neoantigen_rows.to_csv("shared_tumor_neoantigens.csv", index=False)

Neoantigen Peptide
..........C    1
..........L    1
..........P    1
..........R    2
..........S    2
              ..
P........      3
SK.......E     1
T......P.E     1
VT......P      1
VT......P.E    1
Name: Tumor Sample, Length: 87, dtype: int64
Neoantigen Peptide
..........R    2
..........S    2
.........C     2
.........L     2
.........P     2
.........R     2
.........S     2
........C      3
........G      2
........P      4
........R      3
........S      3
........T      3
.......S       2
..R......      2
.C.......      2
.C........     2
.D.......      2
.H.......      4
.P.......      3
.R.......      3
.R........     2
.S.......      3
.S........     2
E........      2
P........      3
Name: Tumor Sample, dtype: int64
     HLA_allele Neoantigen Peptide  Normal Rank Score  Tumor Rank Score  \
0    HLA-A24:02         .........S             7.2538            0.0156   
1    HLA-A24:02          ........S             7.2642            0.0198   
2    HLA-A24:02          .....

Neoantigen-based therapies utilize tumor-specific antigens derived from cancer mutations to stimulate immune responses. 

1). Positive Implications
1.1 Personalized Treatment:
Tailored to an individual’s unique cancer mutations, enhancing efficacy while minimizing off-target effects.

1.2 Reduced Side Effects:
Since neoantigens are tumor-specific, they significantly lower the risk of autoimmune reactions and damage to healthy tissue.

1.3 Synergy with Other Therapies:
Can be combined with immunotherapies such as checkpoint inhibitors to enhance overall treatment effectiveness.

2). Challenges and Limitations
2.1 Complexity and Cost:
Requires extensive genomic analysis, bioinformatics assessment, and personalized manufacturing, increasing both cost and complexity.

2.2 Uncertain Immunogenicity:
Not all neoantigens elicit strong immune responses, making reliable identification a challenge.

2.3 Tumor Heterogeneity and Immune Evasion:
Variability within tumors may result in incomplete targeting, allowing some cancer cells to escape immune detection and potentially leading to recurrence.

2.4 Lengthy Development Process:
Personalized production and validation can take weeks or months, limiting accessibility for patients with rapidly progressing cancers.

2.5 Ethical and Accessibility Concerns:
High costs and the need for specialized facilities may restrict availability, exacerbating healthcare disparities.










2) Challenges and Limitations
2.1 Complexity and Cost:
Requires extensive genomic analysis, bioinformatics assessment, and personalized manufacturing, increasing both cost and complexity.

2.2 Uncertain Immunogenicity:
Not all neoantigens elicit strong immune responses, making reliable identification a challenge.

2.3 Tumor Heterogeneity and Immune Evasion:
Variability within tumors may result in incomplete targeting, allowing some cancer cells to escape immune detection and potentially leading to recurrence.
